src/big/big_decimal.cr

```diff
@@ -0,0 +1,249 @@
+require "big_int"
+
+# A `BigDecimal` represents arbitrary precision decimals.
+#
+# It is internally represented by a pair of `BigInt` and `UInt64`: value and scale.
+# Value contains the actual value, and scale tells the decimal point place.
+# e.g. value=1234, scale=2 => 12.34
+#
+# The general idea and some of the arithmetic algorithms were adapted from
+# the MIT/APACHE -licensed https://github.com/akubera/bigdecimal-rs
+
+class InvalidBigDecimalException < Exception
+end
+
+struct BigDecimal
+  ZERO = BigInt.new(0)
+  TEN  = BigInt.new(10)
+  @@max_div_iterations = 100_u64
+
+  include Comparable(BigDecimal)
+  include Comparable(Int)
+  include Comparable(String)
+
+  private property value : BigInt
+  private property scale : UInt64
+  getter value, scale
+
+  # Sets maximum iterations used in division operation. This implicitly
+  # defines the maximum precision of divisions in cases where the
+  # division is not exact.
+  def self.set_max_div_iterations(new_max_div : Int)
+    @@max_div_iterations = new_max_div.to_u64
+  end
+
+  # Creates a new `BigDecimal` from a `String`.
+  #
+  # Allows only valid number strings with an optional negative sign.
+  def initialize(str : String)
+    # disallow every non-valid string
+    if str !~ /^-?[0-9]+(\.[0-9]+)?$/
```

Comment:
Actually here's a better way: 

```cr
reader = Char::Reader.new(str)
value = BigInteger.new(0)
scale = 0_u64 

if  reader.current_char == '-' 
  negative = true 
  reader.next_char
end 

has_decimal = false
until (char = reader.current_char) == '\0' 
  if char == '.' 
    raise ArgumentError.new if has_decimal 
    has_decimal = true 
    next 
  end
    
  int = char.to_i?
  raise ArgumentError.new unless int 

  value += int 
  scale += 1
  reader.next_char
end 

value = -value if negative 
```

I wrote this on my phone while on holiday so please fix all the mistakes!

Reply:
Just two errors, BigInteger => BigInt and missing reader.next_char in the until/if block :) 👍 

I benchmarked this against the regexp initializer, and it seemed to be a tiny bit faster with slower numbers and quite a lot slower with big numbers than the regexp-based initializer. 

```
                                        initialize 123.456 regexp   1.73  (578.12ms) (± 3.75%)  2.52× slower
                                        initialize 123.456 reader   1.95  (512.82ms) (± 3.77%)  2.24× slower
                                            initialize 123 regexp   3.82  (261.56ms) (± 2.27%)  1.14× slower
                                            initialize 123 reader   4.36  (229.35ms) (± 3.30%)       fastest
 initialize 0.12389127391827094872903841729038471290347808 regexp   1.37  (731.63ms) (± 1.15%)  3.19× slower
 initialize 0.12389127391827094872903841729038471290347808 reader   0.27  (  3.71s ) (± 0.05%) 16.20× slower
initialize 12387192378912738917128930.11729038471290347808 regexp   1.35  (739.18ms) (± 0.85%)  3.22× slower
initialize 12387192378912738917128930.11729038471290347808 reader   0.26  (  3.85s ) (± 0.38%) 16.80× slower
```

Do you still think this is a good idea?

Reply:
Well at first glance I would think those benchmark results are highly suspect. I'll benchmark it myself and see what happened.

Reply:
OK, I benchmarked it yourself, and you're right: it is slower. I suspect that the problem is the bigdecimal operations of `* 10` and `+ int` are slowing it down. Whereas with your solution, it's fast because it passes only the string to libgmp, which uses a fast path to parse it and turn it into a bigdecimal.

So I decided to take the best parts of both (my solution doesn't use regex, your solution uses libgmp as little as possible) and write a third solution, [here](https://github.com/RX14/crystal/commit/f6ee274674f7ca7b1bec824a4b08a0dffa61bf0e).

Benchmark code [here](https://gist.github.com/RX14/1429c54b7ab2f97578cbdf090f46966b):
```cr
Testing: 123.456
new2   4.23M (236.22ns) (± 4.98%)       fastest
 new   1.08M (929.53ns) (± 7.19%)  3.94× slower
 old   1.64M (608.84ns) (± 5.58%)  2.58× slower
Testing: 123
new2   8.81M (113.48ns) (± 7.95%)       fastest
 new   2.46M (405.82ns) (± 4.22%)  3.58× slower
 old   3.63M ( 275.8ns) (± 4.67%)  2.43× slower
Testing: 0.12389127391827094872903841729038471290347808
new2   2.56M (390.89ns) (± 2.44%)       fastest
 new 142.17k (  7.03µs) (± 4.53%) 17.99× slower
 old   1.22M (816.99ns) (± 7.29%)  2.09× slower
Testing: 12387192378912738917128930.11729038471290347808
new2   2.57M (388.77ns) (± 6.05%)       fastest
 new 134.51k (  7.43µs) (± 4.12%) 19.12× slower
 old   1.25M (799.48ns) (± 5.61%)  2.06× slower
```

My patch does change the semantics a bit: `.2` and `2.` syntaxes now work and yield `0.2` and `2` correctly. This can easilly be changed (it's an extra ` || index == 0 || (index == 1 && str[0] == '-'`), but the code is actually easier and simpler being a bit more permissive, and I think these two syntaxes are quite well-used anyway.